src/20240421_Multi_AprilFinanceUpdate/AaveV2Ethereum_AprilFinanceUpdate_20240421.sol

```diff
@@ -0,0 +1,227 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {IERC20} from 'solidity-utils/contracts/oz-common/interfaces/IERC20.sol';
+import {SafeERC20} from 'solidity-utils/contracts/oz-common/SafeERC20.sol';
+import {MiscEthereum} from 'aave-address-book/MiscEthereum.sol';
+import {AaveV2Ethereum, AaveV2EthereumAssets} from 'aave-address-book/AaveV2Ethereum.sol';
+import {AaveV3Ethereum, AaveV3EthereumAssets} from 'aave-address-book/AaveV3Ethereum.sol';
+import {AaveSwapper} from 'aave-helpers/swaps/AaveSwapper.sol';
+import {IProposalGenericExecutor} from 'aave-helpers/interfaces/IProposalGenericExecutor.sol';
+/**
+ * @title April Finance Update
+ * @author @karpatkey_TokenLogic
+ * - Snapshot: Direct-to-AIP
+ * - Discussion: https://governance.aave.com/t/arfc-april-finance-update/17390
+ */
+contract AaveV2Ethereum_AprilFinanceUpdate_20240421 is IProposalGenericExecutor {
+  using SafeERC20 for IERC20;
+
+  struct TokenToMigrate {
+    address underlying;
+    address aToken;
+    uint256 qty;
+  }
+
+  struct TokenToSwap {
+    address underlying;
+    address aToken;
+    uint256 balance;
+  }
+
+  AaveSwapper public constant SWAPPER = AaveSwapper(MiscEthereum.AAVE_SWAPPER);
+
+  address public constant MILKMAN = 0x11C76AD590ABDFFCD980afEC9ad951B160F02797;
+  address public constant PRICE_CHECKER = 0xe80a1C615F75AFF7Ed8F08c9F21f9d00982D666c;
+  address public constant GHO_USD_FEED = 0x3f12643D3f6f874d39C2a4c9f2Cd6f2DbAC877FC;
+
+  address public constant AGD_MULTISIG = 0x89C51828427F70D77875C6747759fB17Ba10Ceb0;
+  uint256 public constant AGD_GHO_ALLOWANCE = 613_000 ether;
+
+  uint256 public constant USDC_V3_TO_SWAP = 1_400_000e6; // 1.4M
+  uint256 public constant DAI_V3_TO_SWAP = 2_000_000 ether;
+  uint256 public constant USDT_V2_TO_KEEP = 650_000e6; // 650,000
+
+  address public constant BGD_RECIPIENT = 0xb812d0944f8F581DfAA3a93Dda0d22EcEf51A9CF;
+  uint256 public constant USDC_AMOUNT_REIMBURSEMENT = 42_000e6;
+  uint256 public constant USDT_AMOUNT_REIMBURSEMENT = 109_200e6;
+  uint256 public constant LINK_AMOUNT_REIMBURSEMENT = 1640 ether;
+
+  function execute() external {
+    _bgdReimbursements();
+    _agdAllowance();
+    _v2ToV3Migration();
+    _swaps();
+  }
+
+  function _bgdReimbursements() internal {
+    AaveV2Ethereum.COLLECTOR.transfer(
+      AaveV2EthereumAssets.USDC_A_TOKEN,
+      BGD_RECIPIENT,
+      USDC_AMOUNT_REIMBURSEMENT
+    );
+    AaveV2Ethereum.COLLECTOR.transfer(
+      AaveV2EthereumAssets.USDT_A_TOKEN,
+      BGD_RECIPIENT,
+      USDT_AMOUNT_REIMBURSEMENT
+    );
+    AaveV2Ethereum.COLLECTOR.transfer(
+      AaveV2EthereumAssets.LINK_A_TOKEN,
+      BGD_RECIPIENT,
+      LINK_AMOUNT_REIMBURSEMENT
+    );
+  }
+
+  function _agdAllowance() internal {
+    AaveV3Ethereum.COLLECTOR.approve(AaveV2EthereumAssets.USDT_A_TOKEN, AGD_MULTISIG, 0);
+    AaveV3Ethereum.COLLECTOR.approve(
+      AaveV3EthereumAssets.GHO_UNDERLYING,
+      AGD_MULTISIG,
+      AGD_GHO_ALLOWANCE
+    );
```

Comment:
@efecarranza I would add an if condition here, to approve the GHO amount only when the allowance of aUSDT to AGD is 612.9k as if claimed mistakenly before the payload execution, the transfer could be done twice.

Reply:
@brotherlymite sorry say that again, i don't think i understood.

Reply:
```
if (USDT_A_TOKEN.allowance(COLLECTOR, AGD_MULTISIG) >= ~612.9e6) {
  AaveV3Ethereum.COLLECTOR.approve(
      AaveV3EthereumAssets.GHO_UNDERLYING,
      AGD_MULTISIG,
      AGD_GHO_ALLOWANCE
    );
}
```
@efecarranza if the AGD wallet claims aUSDT before proposal execution, we still approve the GHO token. So there is a double approval. So just an extra check.